### PR TITLE
[Core] Adding slip constraints.h

### PR DESCRIPTION
**Pull request by RiccardoRossi**
This PR implements a simple interface for LinearMasterSlaveConstraints from python (basing on the capabilities introduced in #6099, which are slightly improved here )

It also introduced a simple "SlipConstraint" designed to simplify the imposition of "slip" boundary conditions, for example in the case of a structure.

as it is now the test passes (i tried 1000s of repetitions to be sure)
**HOWEVER if the ResidualCriteria is used instead of the DisplacementCriteria the test fails sometimes (when OPENMP>1)**
This points to some sort of race condition in the ResidualCriteria (which SOMETIMES makes the iterations to stop before they are expected to)
I am pretty sure that this is not related to the current PR, nevertheless it is something we need to look into, also because this test is very easy to reproduce. (I'll open a separate issue about that)

### CODE REVIEW

File: kratos/python/add_dofs_to_python.cpp [hunk 1]
```diff
@@ -44,6 +44,12 @@ void  AddDofsToPython(pybind11::module& m)
     .def("Fix", &Dof<double>::FixDof)
     .def("Free", &Dof<double>::FreeDof)
     .def("IsFixed", &Dof<double>::IsFixed)
+    .def("__lt__", [](Dof<double>& dof1, Dof<double>& dof2){return dof1<dof2; })
```
Comment (RiccardoRossi) [inline kratos/python/add_dofs_to_python.cpp:47]: this is just to allow sorting dofs in python, a feature i will need for the next step i am planning

File: kratos/includes/slip_constraint.h [hunk 2]
```diff
@@ -0,0 +1,352 @@
+//    |  /           |
```
Comment (loumalouomega) [inline kratos/includes/slip_constraint.h]: Constraints are supposed to be in a constraint folder, otherwise the includes folder is populated with many files (we should move linear_master_slave_constraint IMO, as it is not the base constraint)

File: kratos/python/add_constraint_to_python.cpp [hunk 3]
```diff
@@ -7,7 +7,7 @@
 //  License:		 BSD License
 //					 Kratos default license: kratos/license.txt
 //
-//  Main authors:    Aditya Ghantasala
+//  Main authors:    Riccardo Rossi
```
Comment (loumalouomega) [inline kratos/python/add_constraint_to_python.cpp:10]: Why removing Aditya?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
I would move the constraint to a constraint folder, in order to avoid overpopulate the includes folder

File: kratos/python/add_dofs_to_python.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/python/add_dofs_to_python.cpp:47]: I thought sorting the dofs was no longer allowed / possible?
@pooyan-dadvand

File: kratos/includes/slip_constraint.h [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+/**
+ * @class SlipConstraint
+ * @ingroup KratosCore
+ * @brief This constructs a constraint which imposes that v * n = 0 
+ * where:
+ *  v is a variable
+ *  n is the unit normal to the node
+ *
+ */
+class SlipConstraint
```
Comment (philbucher) [inline kratos/includes/slip_constraint.h]: please separate int header/source

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        tang_2 = R2 - nR2*normal_2
+        tang_4 = R4 - nR4*normal_4
+
+        self.assertTrue(tang_2.norm_2() < 1e-12)
+        self.assertTrue(tang_4.norm_2() < 1e-12)
+
+        self.assertEqual(mp.ProcessInfo[KratosMultiphysics.NL_ITERATION_NUMBER], 5) #4 if using Residual Criteria
+
+    def __post_process(self, main_model_part):
+        from gid_output_process import GiDOutputProcess
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]: please either remove or fix, this does not work

File: kratos/python/add_constraint_to_python.cpp (hunk 3 above)
Comment (philbucher) [inline kratos/python/add_constraint_to_python.cpp:10]:
```suggestion
//  Main authors:    Aditya Ghantasala, Riccardo Rossi
```

**Review by philbucher — CHANGES_REQUESTED** [review]
found some things while browsing through

**Comment by philbucher** [comment]
check codacy there are some relevant things

btw please set your editor to auto-remove the trailing whitespaces

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py (hunk 5 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]: removed, it was a leftover

File: kratos/python/add_constraint_to_python.cpp (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/python/add_constraint_to_python.cpp:10]: I created this file for new in #6099 Aditya is there only because i copied a file in the same directory which happend to have him as author and forgot changing.

File: kratos/python/add_dofs_to_python.cpp (hunk 1 above)
Comment (RiccardoRossi) [reply kratos/python/add_dofs_to_python.cpp:47]: sorting dofs is definitely possible, we are doing this all the time in the strategy to Unique the dof list.

i added this since i plan to provide an interface to provide dofs, that will require me to be able to compare dofs, and eventually to sort lists of dofs. If you see i am doing nothing else than providing the comparison that is implemented in c++

File: kratos/includes/slip_constraint.h (hunk 2 above)
Comment (RiccardoRossi) [reply kratos/includes/slip_constraint.h]: i am moving, but only the new file. 
moving the others should be done in a different PR, without any other modifications (other than what is needed)

File: kratos/includes/slip_constraint.h (hunk 4 above)
Comment (RiccardoRossi) [reply kratos/includes/slip_constraint.h]: i tend to disagree about this. 

this file is a very thin wrapper on the top of the LinearMasterSlaveConstraint, which essentially makes it easy to use for applying a slip constraint.  If you see the only relevant change is in the constructor.
I don't dislike everything being in the same file.

anyhow, if you are really convinced about doing the splitting, i'll do without insisting more.

**Review by loumalouomega — DISMISSED** [review]

File: kratos/constraints/slip_constraint.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
+//    |  /           |
```
Comment (RiccardoRossi) [inline kratos/constraints/slip_constraint.cpp:1]: ok, i did the splitting since it was actually less work to do it then to discuss further. (also admittedly, my concern was only about a coding taste, i see the advantage of having it splitted)

**Review by RiccardoRossi — COMMENTED** [review]
ok, in principle i did the requested changes. 

thx @loumalouomega for approving. btw, take a look to how the slip constraint can be imposed, it may be useful for your applications

@philbucher about the sorting, i am only adding the possibility to compare dofs (which internally compares the Id of the node they belong to, and then the ordering of the variables)

**Comment by loumalouomega** [comment]
> thx @loumalouomega for approving. btw, take a look to how the slip constraint can be imposed, it may be useful for your applications
> 

I already looked at the code, and I do basically the same.

**Comment by RiccardoRossi** [comment]
@loumalouomega so you are no longer adding auxiliary dofs?

**Comment by RiccardoRossi** [comment]
btw, codacy does not seem relevant to me in this case...

**Comment by loumalouomega** [comment]
In the contactstructural not, in the other yes...

**Review by loumalouomega — DISMISSED** [review]

File: kratos/constraints/slip_constraint.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    all_dofs.reserve(3);
+    all_dofs.push_back(rNode.pGetDof(rVarX));
+    all_dofs.push_back(rNode.pGetDof(rVarY));
+    all_dofs.push_back(rNode.pGetDof(rVarZ));
+
+    auto n = rNode.FastGetSolutionStepValue(rNormalVar);
+    KRATOS_DEBUG_ERROR_IF(norm_2(n) < std::numeric_limits<double>::epsilon()) << "The norm of the normal vector is zero or almost zero" << std::endl;
+    n /= norm_2(n);
+
+    array_1d<double,3> v;
```
Comment (philbucher) [inline kratos/constraints/slip_constraint.cpp]: what is this? Looks like you are not using it

File: kratos/constraints/slip_constraint.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    mConstantVector.clear();
+
+    DofPointerVectorType all_dofs;
+    all_dofs.reserve(3);
+    all_dofs.push_back(rNode.pGetDof(rVarX));
+    all_dofs.push_back(rNode.pGetDof(rVarY));
+    all_dofs.push_back(rNode.pGetDof(rVarZ));
+
+    auto n = rNode.FastGetSolutionStepValue(rNormalVar);
+    KRATOS_DEBUG_ERROR_IF(norm_2(n) < std::numeric_limits<double>::epsilon()) << "The norm of the normal vector is zero or almost zero" << std::endl;
```
Comment (philbucher) [inline kratos/constraints/slip_constraint.cpp]:
```suggestion
    KRATOS_ERROR_IF(norm_2(n) < std::numeric_limits<double>::epsilon()) << "The norm of the normal vector is zero or almost zero" << std::endl;
```

this check is only done in the constructor

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+    def _apply_BCs(self,mp,A,b):
+        pass
+
+    def _create_strategy(self, mp):
+        #define a minimal newton raphson solver
+        linear_solver = KratosMultiphysics.SkylineLUFactorizationSolver()
+        builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolver(linear_solver)
+        scheme = KratosMultiphysics.ResidualBasedIncrementalUpdateStaticScheme()
+#        convergence_criterion = KratosMultiphysics.ResidualCriteria(1e-13,1e-14) #Parallelism problem, not sure where!
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]:
```suggestion
#        convergence_criterion = KratosMultiphysics.ResidualCriteria(1e-13,1e-14)
```

better remove this, it is solved now

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+import math
+
+def inner_prod(a,b):
+    tmp = 0
+    for i in range(len(a)):
+        tmp += a[i]*b[i]
+    return tmp
+
+class TestLinearConstraints(KratosUnittest.TestCase):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]: I would put this to the core, see https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_linear_master_slave_constraints.py

File: kratos/constraints/slip_constraint.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    mConstantVector.resize(1,false);
+    mConstantVector.clear();
+
+    DofPointerVectorType all_dofs;
+    all_dofs.reserve(3);
+    all_dofs.push_back(rNode.pGetDof(rVarX));
+    all_dofs.push_back(rNode.pGetDof(rVarY));
+    all_dofs.push_back(rNode.pGetDof(rVarZ));
+
+    auto n = rNode.FastGetSolutionStepValue(rNormalVar);
```
Comment (philbucher) [inline kratos/constraints/slip_constraint.cpp]: not sure what you are using it for, but are you sure this should be in the hist-database?
I had the impression that for those "helper" vars like `NORMAL` we are moving towards the non-hist database

**Review by philbucher — COMMENTED** [review]
some comments

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py (hunk 10 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]: even though this feature could be used with any application, this test requires the structural mechanics, so i am not clear it is a good idea to put it in the core, conditionally to the structural mechancis being compiled

File: kratos/constraints/slip_constraint.cpp (hunk 7 above)
Comment (RiccardoRossi) [reply kratos/constraints/slip_constraint.cpp]: removed

File: kratos/constraints/slip_constraint.cpp (hunk 11 above)
Comment (RiccardoRossi) [reply kratos/constraints/slip_constraint.cpp]: i changed the constructor so to pass directly n. this way one can choose what to use explicitly

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        dy = n1.Y - n4.Y
+        normal_4 = KratosMultiphysics.Array3([dy,-dx,0])/math.sqrt(dx**2+dy**2)
+        n4.SetValue(KratosMultiphysics.NORMAL, normal_4) #artificially set the normal
+
+        constraint_4 = KratosMultiphysics.SlipConstraint(4,
+            n4,
+            KratosMultiphysics.DISPLACEMENT_X,
+            KratosMultiphysics.DISPLACEMENT_Y,
+            KratosMultiphysics.DISPLACEMENT_Z,
+            n4.GetValue(KratosMultiphysics.NORMAL),
```
Comment (RiccardoRossi) [inline applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]: here i pass directly "n" so that a user can decide what to do

**Comment by philbucher** [comment]
> even though this feature could be used with any application, this test requires the structural mechanics, so i am not clear it is a good idea to put it in the core, conditionally to the structural mechancis being compiled

@RiccardoRossi the test in the core also needs StructuralMechanics. It does exactly the same basically (well the test itself is different ofc...)
Since this is an important (Core) feature I think it should be in the core, even though it might be skipped

File: applications/StructuralMechanicsApplication/tests/test_linear_constraints.py (hunk 10 above)
Comment (RiccardoRossi) [reply applications/StructuralMechanicsApplication/tests/test_linear_constraints.py]: done

File: kratos/tests/test_linear_constraints.py [hunk 13]
```diff
@@ -0,0 +1,194 @@
+from __future__ import print_function, absolute_import, division
+import KratosMultiphysics
+import KratosMultiphysics.StructuralMechanicsApplication as StructuralMechanicsApplication
```
Comment (philbucher) [inline kratos/tests/test_linear_constraints.py]: this has to be protected, otherwise crashes when Structural is not compiled
```suggestion
```

the protected import is a few lines later

**Review by philbucher — COMMENTED** [review]
I would combine the two test files, otherwise it is confusing IMO
not a blocker though

File: kratos/tests/test_linear_constraints.py (hunk 13 above)
Comment (RiccardoRossi) [reply kratos/tests/test_linear_constraints.py]: yes i wanted to do that but i didn't remove it from above

**Comment by RiccardoRossi** [comment]
> I would combine the two test files, otherwise it is confusing IMO
> not a blocker though

I don't follow what you are asking...

we have here a test for the dof list and one for the linear constraints...

these are two different things...

**Comment by RiccardoRossi** [comment]
I see that the CI/ubuntu is failing however i think this is completely unrelated to this PR.

**Comment by loumalouomega** [comment]
> 
> 
> I see that the CI/ubuntu is failing however i think this is completely unrelated to this PR.

~~~
corrupted size vs. prev_size
Aborted (core dumped)
##[error]Process completed with exit code 134.
~~~

@roigcarlo what could be this?

**Comment by roigcarlo** [comment]
I think that's the test @philbucher is fixing right now isn't it?

**Comment by philbucher** [comment]
I am on it, don't know the source yet, looking since yesterday

**Review by philbucher — APPROVED** [review]
